### ISSUE TITLE
"keep" did not retain newest images

### ISSUE THREAD

**Issue by GAEfan** [GoogleCloudPlatform/gcr-cleaner#89]
Contrary to my expectations, the "_keep_" parameter did not retain the **_newest_** X images.  The remaining images are from random dates.

Is there a regex expression to use in "_tag_filter_any_", so that it keeps the images with the "_latest_" tag?  My tags look something like:

(b095b33c-9b58-aaaa-bbbb-xxxxxxxxxxxxxxx) (latest)
(xnsdfnwe4-wr5f-9b58-bbbb-xxxxxxxxxxxxxxx) 
(kkfgn8jf8j4-q4fg-aaaa-1233-xxxxxxxxxxxxxxx) 
etc.

Thanks for the great package!  My artifact registry had grown into the hundreds of GB.

**Comment by sethvargo** [comment]
Images are kept from their _creation_ date. 

1. Why query did you run?
2. What is the output of the command with `GCRCLEANER_LOG=debug`?

**Comment by sethvargo** [comment]
Do you have `docker` installed?

If you're running it via Cloud Run, you'll need to set the environment variable to `GCRCLEANER_LOG` to `debug`

**Comment by GAEfan** [comment]
Yes, Docker is installed.  I tried with it running and not

Attached is a log file in JSON format.  For some reason, Google's CSV output is not formatted correctly.
[downloaded-logs-20220718-153815.csv](https://github.com/GoogleCloudPlatform/gcr-cleaner/files/9136339/downloaded-logs-20220718-153815.csv)

**Comment by sethvargo** [comment]
Hi @GAEfan 

It's not a CSV, it's a JSON document. What version of gcr-cleaner are you running?

Images are deleted by _created_ date, not _updated_ date.

**Comment by GAEfan** [comment]
Yes, I downloaded in JSON format. What I meant was that when I download as CSV, Google corrupts the format. So I used JSON instead.

I do not see the version listed.  But I installed it using the instructions here just yesterday, so it should be current.  I got the image from `us-docker.pkg.dev/gcr-cleaner/gcr-cleaner/gcr-cleaner` per your instructions.

Yes, I want it to sort by _creation date_, and delete the oldest.  But, it looks like is is not sorting by _creation date_ before deleting.

**Comment by GAEfan** [comment]
Yes, that was what I was expecting.  But, when I ran it, this:

<img width="924" alt="Screen Shot 2022-07-18 at 3 35 12 PM_edit" src="https://user-images.githubusercontent.com/2022696/179782241-3f3d2481-be32-46e1-947b-ff279bb8f018.png">

turned into this:

<img width="914" alt="Screen Shot 2022-07-18 at 3 36 34 PM_edit" src="https://user-images.githubusercontent.com/2022696/179782263-2441576a-e1af-4698-85c4-107406c24f66.png">


Just by random chance, the _latest_ version still remained.

**Comment by sethvargo** [comment]
For the `ttl-18h` repo, the first 3 repos it found were:

- `sha256:ba591f4b476a3450e3b43d3c22a036aa71658a21ac1ed2511b101f5867f8b64e` (`1f08bfef-3716-44ea-8cf6-fc9ad76ad16e`)
- `sha256:ec692229b75f3ed581434245f9c5db574c088e4ab0dbe2a19031ed1e731ae483` (`a69b86cd-f0e6-434f-8dc4-669f7e9d3007`)
- `sha256:022f1544b5d6c381b0ea5896a1e7240f37a00622dda854c1dd78c363c349714e` (`49acb344-f755-45b5-bd37-71bffd6aaf57`)

After that, it deleted:

- `sha256:16098d13ba6fb56180a960eac990e052281ab0d45a32eaa1849d15966594df38` (`f0c85afd-b35a-4ca5-bd9d-916cc5408fdb`)
- `sha256:1caadfb34077cdc73760db3e48c41883c4d71ccf91789f6c6bd84133a2db428c` (`856517be-b2bc-4cc4-b87a-67ec5df4a345`)
- `sha256:8958ac4c19acad39f479729fbd554713cc692dc10d15f7e313a4c2712d63b968` (`433d5ede-25f0-4a25-8d5c-99e7ca2bcc98`)
- `sha256:9a7a645eccf541b86e44e3d27b4e3165dffef17f3c63bc377dfaba3e2ded07ed` (`0c2f81a5-9a10-4798-8b24-0a61f56fb720`, `latest`)

Then it moved on to `ttl-7d`, the first 3 repos it found were:

- `sha256:066095b8a841f1af76da96d63e3481c9f056a4cc1f2bbaab50f24b1cba0b528c` (`b4085970-83fa-4981-b712-d4874cdd4638`, `latest`)
- `sha256:4c0c6576f4aef77382533ec4f5aaf73a46406cfb3d809c0fb9f1b916c66dd39f` (`9b012d86-8e3c-4ead-89d5-6ad2a738bc03`)
- `sha256:58d71ac7daa6c2dd181e66d7279a118faa328804137c463b103839905d15b0ba` (`d51a0e6f-1b9e-4a47-83fa-4f96545fc723`)

Then it deleted:

- `sha256:6bdcf42f466176bb801f81fc3000d324c1b0ee0f25f21d151ce2d906bdde7288` (`f9b9e405-a872-4c7d-94b0-49f5e798737d`)
- `sha256:6edf7fd662a6ab1ab7649615408cf5f62519a99287f81c2b01eff3c3118066b4` (`6bb12ed6-a33f-4679-b00d-e65dae5ab008`)
- `sha256:6f49c07833ac20cb3d4781c09fc61974f5a8e87d804e767e67174333ba99d77b` (`2b7e40aa-4dc8-441e-a3dc-080dd8aaa2fe`)
- `sha256:ff1464efd9b145cb1a19ef3e6dd87e0a4867dc203bc892ecc1bb81e2e6bfcf2a` (`28c97fff-47dc-4782-8590-2ba0b6aaeb97`)

I opened https://github.com/GoogleCloudPlatform/gcr-cleaner/pull/90 which would add more detailed timestamp information to see if that is the _right_ order of operations.

**Comment by GAEfan** [comment]
Great!  I look forward to testing that. The images I posted above were from `ttl-7d`. The first 3 it found were not in proper chron order.

**Comment by GAEfan** [comment]
OK, getting somewhere... The _created_ tags are not being set properly. In my case, _uploaded_ is the timestamp for when I ran `cloud deploy`, and would be the same as _created_, but not sure that is the case for everyone.
```
"uploaded": "2022-07-18T22:14:33Z",
"created": "1980-01-01T00:00:01Z",
```

**Comment by sethvargo** [comment]
Hmm - that feels like an Artifact Registry bug.

**Comment by GAEfan** [comment]
Yeah, that seems right.  This seems to be intentional on Google's part:

https://github.com/GoogleCloudPlatform/buildpacks/issues/21

Can you add a secondary sort on _uploaded_ date?

**Comment by GAEfan** [comment]
Great, thanks! v0.8.0 seems to be working fine.  Thanks for the great package and prompt responses!
